src/main/java/com/puppycrawl/tools/checkstyle/site/ViolationMessagesMacro.java

```diff
@@ -0,0 +1,274 @@
+
+        // deep scan class through hierarchy
+        final Class<?> superModule = clss.getSuperclass();
+
+        if (superModule != null && shouldScanDeepClassForMessages(superModule)) {
+            checkstyleMessages.addAll(getCheckMessageKeyFields(superModule));
+        }
+
+        // special cases that require additional classes
+        try {
+            if (clss == RegexpMultilineCheck.class) {
+                checkstyleMessages.addAll(getCheckMessageKeyFields(Class
+                    .forName("com.puppycrawl.tools.checkstyle.checks.regexp.MultilineDetector")));
+            }
+            else if (clss == RegexpSinglelineCheck.class
+                    || clss == RegexpSinglelineJavaCheck.class) {
+                checkstyleMessages.addAll(getCheckMessageKeyFields(Class
+                    .forName("com.puppycrawl.tools.checkstyle.checks.regexp.SinglelineDetector")));
+            }
+        }
+        catch (ClassNotFoundException ex) {
+            throw new MacroExecutionException("Couldn't find class", ex);
+        }
+
+        return checkstyleMessages;
+    }
+
+    /**
+     * Should the class be deep scanned for messages.
+     *
+     * @param superModule The class to examine.
+     * @return {@code true} if the class should be deeply scanned.
+     */
+    private static boolean shouldScanDeepClassForMessages(Class<?> superModule) {
+        return superModule == AbstractNameCheck.class
+                || superModule == AbstractAccessControlNameCheck.class
+                || superModule == AbstractParenPadCheck.class
+                || superModule == AbstractSuperCheck.class;
+    }
+
+    /**
+     * Returns the message keys for the given fields.
+     *
+     * @param messageKeyFields the fields.
+     * @param instance the instance of the module.
+     * @return the message keys.
+     * @throws MacroExecutionException if the message keys could not be retrieved.
+     */
+    private static Set<String> getMessageKeys(Set<Field> messageKeyFields, Object instance)
+            throws MacroExecutionException {
+        final Set<String> messageKeys = new TreeSet<>();
+        for (Field field : messageKeyFields) {
+            messageKeys.add(getFieldValue(field, instance));
+        }
+        return messageKeys;
+    }
+
+    /**
+     * Returns the value of the given field.
+     *
+     * @param field the field.
+     * @param instance the instance of the module.
+     * @return the value of the field.
+     * @throws MacroExecutionException if the value could not be retrieved.
+     */
+    private static String getFieldValue(Field field, Object instance)
+            throws MacroExecutionException {
+        try {
+            // required for package/private classes
+            field.trySetAccessible();
+            return field.get(instance).toString();
+        }
+        catch (IllegalAccessException ex) {
+            throw new MacroExecutionException("Couldn't get field value", ex);
+        }
+    }
```

Comment:
All this is a literal copy of https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/CheckUtil.java.

But I have no access to it.

Reply:
Hm, this brings up a good point. Should this be considered `internal`? @romani @rnveach

Reply:
it is Util in test area of code, I am ok make all methods that we need public

Reply:
They are public, the thing is, I can not import code from `test` in `main` :/

Reply:
macros code will go to separate project, as we should not export by jar such doxia site classes :).
please keep them as copy, but you can make SiteUtil.java to make copy-paste more explicit and declarative.